[ML] Improve hard_limit audit message

edsavage opened the pull request:
Improve the hard_limit memory audit message by reporting how many bytes
over the configured memory limit the job was at the point of the last
allocation failure.

Previously the model memory usage was reported, however this was
inaccurate and hence of limited use -  primarily because the total
memory used by the model can decrease significantly after the models'
status is changed to hard_limit but before the model size stats are
reported from autodetect to ES.

While this PR contains the changes to the format of the hard_limit audit
message it is dependent on modifications to the ml-cpp backend to
send additional data fields in the model size stats message. These
changes will follow in a subsequent PR. It is worth noting that this PR
must be merged prior to the ml-cpp one, to keep CI tests happy.

Relates #38034

edsavage committed:
[ML] Improve hard_limit audit message

Improve the hard_limit memory audit message by reporting how many bytes
over the configured memory limit the job was at the point of the last
allocation failure.

Previously the model memory usage was reported, however this was
inaccurate and hence of limited use -  primarily because the total
memory used by the model can decrease significantly after the models
status is changed to hard_limit but before the model size stats are
reported from autodetect to ES.

While this PR contains the changes to the format of the hard_limit audit
message it is dependent on modifications to the ml-cpp backend to
send additional data fields in the model size stats message. These
changes will follow in a subsequent PR. It is worth noting that this PR
must be merged prior to the ml-cpp one, to keep CI tests happy.

Relates #38034

elasticmachine commented:
Pinging @elastic/ml-core

droberts195 commented on client/rest-high-level/src/main/java/org/elasticsearch/client/ml/job/process/ModelSizeStats.java:
```diff
@@ -148,6 +162,14 @@ public long getModelBytes() {
         return modelBytes;
     }
 
+    public long getModelBytesExceeded() {
```
The return value needs to be `Long`, otherwise a user could get an NPE.

droberts195 commented on client/rest-high-level/src/main/java/org/elasticsearch/client/ml/job/process/ModelSizeStats.java:
```diff
@@ -148,6 +162,14 @@ public long getModelBytes() {
+        return modelBytesExceeded;
+    }
+
+    public long getModelBytesMemoryLimit() {
```
The return value needs to be `Long`, otherwise a user could get an NPE.

droberts195 commented on client/rest-high-level/src/main/java/org/elasticsearch/client/ml/job/process/ModelSizeStats.java:
```diff
@@ -219,6 +242,8 @@ public boolean equals(Object other) {
         private final String jobId;
         private long modelBytes;
+        private long modelBytesExceeded;
+        private long modelBytesMemoryLimit;
```
These two need to be `Long`, otherwise they'll default to `0` and zeroes will then propagate into objects that should really contain `null`s.

droberts195 commented on client/rest-high-level/src/main/java/org/elasticsearch/client/ml/job/process/ModelSizeStats.java:
```diff
@@ -47,6 +47,8 @@
      */
     public static final ParseField MODEL_BYTES_FIELD = new ParseField("model_bytes");
+    public static final ParseField MODEL_BYTES_EXCEEDED_FIELD = new ParseField("model_bytes_exceeded");
+    public static final ParseField MODEL_BYTES_MEMORY_LIMIT_FIELD = new ParseField("model_bytes_memory_limit");
```
This is copied from `analysis_limits.model_memory_limit`, so I think it would best to call this `model_memory_limit`.  That will make it clearer that it refers to the same thing as the `model_memory_limit` from the job config.

droberts195 replied:
On reflection scrub the above comment. `analysis_limits.model_memory_limit` is megabytes whereas this one is bytes so it makes sense to have a slightly different name.

droberts195 commented on client/rest-high-level/src/test/java/org/elasticsearch/client/ml/job/process/ModelSizeStatsTests.java:
```diff
@@ -31,6 +31,8 @@
         ModelSizeStats stats = new ModelSizeStats.Builder("foo").build();
         assertEquals(0, stats.getModelBytes());
+        assertEquals(0, stats.getModelBytesExceeded());
+        assertEquals(0, stats.getModelBytesMemoryLimit());
```
These two will change to `null` if the other changes I recommended are made.  So that makes the values set by the default constructed builder inconsistent.  But this is better than filling in values that didn't really exist in a JSON document.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/process/autodetect/state/ModelSizeStats.java:
```diff
@@ -125,6 +135,16 @@ private ModelSizeStats(String jobId, long modelBytes, long totalByFieldCount, lo
+        if (in.getVersion().onOrAfter(Version.V_7_2_0)) {
+            modelBytesExceeded = in.readOptionalLong();
+        } else {
+            modelBytesExceeded = 0L;
```
I think it should be `null`, so we remember that the field didn't exist.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/process/autodetect/state/ModelSizeStats.java:
```diff
@@ -125,6 +135,16 @@ private ModelSizeStats(String jobId, long modelBytes, long totalByFieldCount, lo
+        if (in.getVersion().onOrAfter(Version.V_7_2_0)) {
+            modelBytesMemoryLimit = in.readOptionalLong();
+        } else {
+            modelBytesMemoryLimit = 0L;
```
I think it should be `null`, so we remember that the field didn't exist.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/process/autodetect/state/ModelSizeStats.java:
```diff
@@ -192,6 +224,14 @@ public long getModelBytes() {
         return modelBytes;
     }
 
+    public long getModelBytesExceeded() {
```
The return value needs to be `Long`, otherwise a user could get an NPE.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/process/autodetect/state/ModelSizeStats.java:
```diff
@@ -192,6 +224,14 @@ public long getModelBytes() {
+        return modelBytesExceeded;
+    }
+
+    public long getModelBytesMemoryLimit() {
```
The return value needs to be `Long`, otherwise a user could get an NPE.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/process/autodetect/state/ModelSizeStats.java:
```diff
@@ -262,6 +304,8 @@ public boolean equals(Object other) {
         private final String jobId;
         private long modelBytes;
+        private long modelBytesExceeded;
+        private long modelBytesMemoryLimit;
```
These two need to be `Long`, otherwise they'll default to `0` and zeroes will then propagate into objects that should really contain `null`s.

droberts195 commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/job/process/autodetect/state/ModelSizeStatsTests.java:
```diff
@@ -22,6 +22,8 @@
         ModelSizeStats stats = new ModelSizeStats.Builder("foo").build();
         assertEquals(0, stats.getModelBytes());
+        assertEquals(0, stats.getModelBytesExceeded());
+        assertEquals(0, stats.getModelBytesMemoryLimit());
```
These two will change to `null` if the other changes I recommended are made.  So that makes the values set by the default constructed builder inconsistent.  But this is better than filling in values that didn't really exist in a JSON document.

edsavage committed:
Attending to code review comments

edsavage committed:
Merge branch 'master' of github.com:elastic/elasticsearch into improve_hard_limit_audit_message

droberts195 commented on client/rest-high-level/src/test/java/org/elasticsearch/client/ml/job/process/ModelSizeStatsTests.java:
```diff
@@ -31,6 +31,8 @@
         ModelSizeStats stats = new ModelSizeStats.Builder("foo").build();
         assertEquals(0, stats.getModelBytes());
+        assertEquals(null, stats.getModelBytesExceeded());
+        assertEquals(null, stats.getModelBytesMemoryLimit());
```
You can use `assertNull` for these two.

edsavage committed:
Backwards compatibility fix.

Use an appropriate hard_limit audit message when model size stats
originate from a version prior to 7.2

edsavage committed:
Further tidy up of test code

droberts195 reviewed (approved):
LGTM
